gossip/src/duplicate_shred_handler.rs

```diff
@@ -139,21 +170,59 @@ impl DuplicateShredHandler {
                 return false;
             }
         }
-        // Also skip frozen slots or slots with an older proof than me.
+        // Also skip frozen slots or slots with a different proof than me.
         match self.chunk_map.get(&slot) {
-            Some(SlotStatus::Frozen) => {
-                return false;
-            }
-            Some(SlotStatus::UnfinishedProof(slot_map)) => {
-                if let Some(proof_chunkmap) = slot_map.peek(&data.from) {
-                    if proof_chunkmap.wallclock < data.wallclock {
-                        return false;
+            Some(SlotStatus::Frozen) => false,
+            Some(SlotStatus::UnfinishedProof(slot_map)) => match slot_map.get(&data.from) {
+                None => self.check_has_enough_stake_and_cleanup(&slot, &data.from),
+                Some(proof_chunkmap) => proof_chunkmap.wallclock == data.wallclock,
+            },
+            None => true,
+        }
+    }
+
+    fn check_has_enough_stake_and_cleanup(&mut self, slot: &Slot, newkey: &Pubkey) -> bool {
+        match self.chunk_map.get_mut(slot) {
+            None => true,
```

Comment:
so if this is a new `slot` we always insert the entry even if we are at capacity?

Reply:
Current the capacity is calculated per slot, of course, we do limit number of slots each pubkey can insert.